### PR TITLE
#164797166 Sets up roles and access control

### PR DESCRIPTION
**Pull request by igbominadeveloper**
#### What does this PR do?
sets up roles and access control

#### Description of Task to be completed?
- [x] add role column to users table
- [x] setup middleware to lock routes to admin only
- [x] setup endpoints to grant and revoke admin access

#### How should this be manually tested?
- Clone the repo,
- Run ```npm install``` to install all dependencies,
- Create .env file using the .env.example as a guide and enter you environment variables
- Run `sequelize db:create`  
- Run `sequelize db:migrate` 
- Run `sequelize db:seed:all`   
- Run npm run start:dev,
- send a POST request to `localhost:3000/api/v1/auth/login` using `igbominadeveloper@ah.com` and password `password1`
- On Postman, send a GET request to localhost:3000/api/v1/roles
- You should get the screenshot below

#### Any background context you want to provide?
Some routes should be accessible to an admin only

#### What are the relevant pivotal tracker stories? (if applicable)
[#164797166](https://www.pivotaltracker.com/story/show/164797166)

#### Screenshots
Screenshot 2019-04-14 at 6 02 20 PM

Roles and access control

### CODE REVIEW

File: src/db/models/Article.js [hunk 1]
```diff
@@ -32,11 +32,11 @@ export default (sequelize, DataTypes) => {
       subscriptionType: {
         type: DataTypes.ENUM('free', 'premium'),
         field: 'subscription_type',
-        default: 'free',
+        defaultValue: 'free',
       },
       status: {
         type: DataTypes.ENUM('published', 'draft', 'trashed', 'banned'),
```
Comment (Japhethca) [inline src/db/models/Article.js]: Let create the enum values as constants in util/constants as 
```javascript
export const PUBLISHED = 'PUBLISHED'; // etc
```

Comment (NedyUdombat) [reply src/db/models/Article.js]: This has been resolved
It has been exported into 
```javascript
src/routes/utils/constants.js
```

File: src/routes/controllers/roles.controller.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * Fetch all users and roles
+ *
+ * @param {Object} req express request object
+ * @param {Object} res express response object
+ * @returns {*} success response
+ * @throws {*} error if database error
+ */
+export async function getAll(req, res) {
```
Comment (Japhethca) [inline src/routes/controllers/roles.controller.js]: lets rename this controller to `getAllRoles`

File: src/routes/controllers/roles.controller.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+ * Make a user an admin
+ *
+ * @param {Object} req express request object
+ * @param {Object} res express response object
+ * @returns {*} success response
+ * @throws {*} error if database error
+ */
+export async function makeAdmin(req, res) {
+  const { email } = req.body;
+  const role = ADMIN;
```
Comment (Japhethca) [inline src/routes/controllers/roles.controller.js]: we can remove this variable and use the admin constant instead.

File: src/routes/controllers/roles.controller.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * Make a user an admin
+ *
+ * @param {Object} req express request object
+ * @param {Object} res express response object
+ * @returns {*} success response
+ * @throws {*} error if database error
+ */
+export async function makeAdmin(req, res) {
```
Comment (Japhethca) [inline src/routes/controllers/roles.controller.js]: we can rename this controller to `updateUserRole` and make handle updating of user roles not just admin role but any role.

File: src/routes/controllers/roles.controller.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * Revoke a users' admin access
+ *
+ * @param {Object} req express request object
+ * @param {Object} res express response object
+ * @returns {*} success response
+ * @throws {*} error if database error
+ */
+export async function revokeAdmin(req, res) {
```
Comment (Japhethca) [inline src/routes/controllers/roles.controller.js]: Given the above comment, we will not need this controller.

File: src/routes/index.js [hunk 6]
```diff
@@ -21,4 +22,6 @@ router.use('/profiles', userRouter);
 router.use('/profiles', followRouter);
 router.use('/users', userRouter);
 
+router.use('/roles', rolesRouter);
```
Comment (Japhethca) [inline src/routes/index.js]: This route should be tied to the users route; `/users/roles`.

File: src/routes/roles.route.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+ *       400:
+ *         description: Bad request.
+ *       401:
+ *         description: Authorization information is missing or invalid.
+ *       403:
+ *        description: Access Forbidden.
+ *       5XX:
+ *        description: Unexpected error.
+ */
+rolesRouter.patch(
```
Comment (Japhethca) [inline src/routes/roles.route.js]: this endpoint should be `/:username`

File: src/routes/roles.route.js [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
```
Comment (Japhethca) [inline src/routes/roles.route.js]: we can remove this, if we make the changes above.

**Review by Japhethca — CHANGES_REQUESTED** [review]
Just few comments to address.

File: src/routes/controllers/roles.controller.js (hunk 2 above)
Comment (NedyUdombat) [reply src/routes/controllers/roles.controller.js]: this has been renamed

File: src/routes/controllers/roles.controller.js (hunk 4 above)
Comment (NedyUdombat) [reply src/routes/controllers/roles.controller.js]: this has been implemented

File: src/routes/controllers/roles.controller.js (hunk 3 above)
Comment (NedyUdombat) [reply src/routes/controllers/roles.controller.js]: This has been resolved

File: src/routes/controllers/roles.controller.js (hunk 5 above)
Comment (NedyUdombat) [reply src/routes/controllers/roles.controller.js]: this controller has been deleted

File: src/routes/index.js (hunk 6 above)
Comment (NedyUdombat) [reply src/routes/index.js]: feedback implemented

File: src/routes/roles.route.js (hunk 8 above)
Comment (NedyUdombat) [reply src/routes/roles.route.js]: this has been remvoed

File: src/routes/roles.route.js (hunk 7 above)
Comment (NedyUdombat) [reply src/routes/roles.route.js]: it has also bee changed to a put

File: src/tests/routes/controllers/roles.test.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+} from '../../db/mockdata/userdata';
+import { AUTHOR, ADMIN } from '../../../utils/constants';
+
+chai.use(chaiHttp);
+const { expect } = chai;
+const rolesEndpoint = '/api/v1/users/roles';
+let userToken;
+let adminToken;
+
+const userEmail = {
```
[the remark this answers is not in the document]
Comment (NedyUdombat) [reply src/tests/routes/controllers/roles.test.js]: this variable has been removed

File: src/tests/routes/controllers/roles.test.js [hunk 10]
```diff
@@ -0,0 +1,140 @@
+import chai from 'chai';
+import chaiHttp from 'chai-http';
+
+import app from '../../../server';
+import {
+  adminCredentials,
+  authorCredentials,
+} from '../../db/mockdata/userdata';
+import { AUTHOR, ADMIN } from '../../../utils/constants';
```
[the remark this answers is not in the document]
Comment (NedyUdombat) [reply src/tests/routes/controllers/roles.test.js]: This constants have been removed

File: package.json [hunk 11]
```diff
@@ -42,7 +42,7 @@
     "check-staged": "lint-staged",
     "prettier": "prettier --write '**/*.{js,jsx}'",
     "pretest": "NODE_ENV=test npm run lint && npm run prettier && npm run db:migrate:testdb && npm run seedall",
-    "test": "NODE_ENV=test nyc mocha --recursive src/tests --require @babel/register --exit",
+    "test": "NODE_ENV=test nyc mocha --recursive src/tests --require @babel/register --exit --no-timeouts",
```
Comment (Japhethca) [inline package.json]: why did we add `--no-timeouts` option to our test?

Comment (NedyUdombat) [reply package.json]: This may have been a mistake I made while rebasing, but it has been resolved now

**Review by Japhethca — DISMISSED** [review]

**Review by Japhethca — APPROVED** [review]

**Pull request merged**